crates/gpui/src/tab_stop.rs

```diff
@@ -32,38 +35,58 @@ impl TabHandles {
         self.handles.clear();
     }
 
-    fn current_index(&self, focused_id: Option<&FocusId>) -> usize {
+    pub(crate) fn with_group(&self, focused_id: Option<&FocusId>) -> Vec<FocusHandle> {
+        if let Some(focused_id) = focused_id {
+            if let Some(handle) = self.handles.iter().find(|h| &h.id == focused_id) {
+                return self
+                    .handles
+                    .iter()
+                    .filter(|h| h.tab_group == handle.tab_group)
+                    .cloned()
+                    .collect();
+            }
+        }
+
         self.handles
             .iter()
-            .position(|h| Some(&h.id) == focused_id)
-            .unwrap_or_default()
+            .filter(|h| h.tab_group.is_none())
+            .cloned()
+            .collect()
     }
 
     pub(crate) fn next(&self, focused_id: Option<&FocusId>) -> Option<FocusHandle> {
-        let ix = self.current_index(focused_id);
+        let group_handles = self.with_group(focused_id);
+        let ix = group_handles
+            .iter()
+            .position(|h| Some(&h.id) == focused_id)
+            .unwrap_or_default();
 
         let mut next_ix = ix + 1;
-        if next_ix + 1 > self.handles.len() {
+        if next_ix >= group_handles.len() {
             next_ix = 0;
         }
 
-        if let Some(next_handle) = self.handles.get(next_ix) {
+        if let Some(next_handle) = group_handles.get(next_ix) {
             Some(next_handle.clone())
         } else {
             None
         }
     }
 
     pub(crate) fn prev(&self, focused_id: Option<&FocusId>) -> Option<FocusHandle> {
-        let ix = self.current_index(focused_id);
-        let prev_ix;
-        if ix == 0 {
-            prev_ix = self.handles.len().saturating_sub(1);
+        let group_handles = self.with_group(focused_id);
+        let ix = group_handles
+            .iter()
+            .position(|h| Some(&h.id) == focused_id)
+            .unwrap_or_default();
+
+        let prev_ix = if ix == 0 {
+            group_handles.len().saturating_sub(1)
         } else {
-            prev_ix = ix.saturating_sub(1);
-        }
+            ix.saturating_sub(1)
+        };
 
-        if let Some(prev_handle) = self.handles.get(prev_ix) {
+        if let Some(prev_handle) = group_handles.get(prev_ix) {
```

Comment:
Here as well, `.cloned()` but remove the whole `if` block

Reply:
Thanks